### DISCUSSION TITLE
Drag`n`Drop from Avalonia app to OS file system (Desktop, etc)

### DISCUSSION THREAD

**Discussion by followLemmi** [AvaloniaUI/Avalonia#15042] [Q&A] [answered]
Is there any possibility in Avalonia to drag and drop a file from the Avalonia application to the Windows desktop?

**Comment by KrzysztofDusko** [comment]
Avalonia -> filesytem not opposite ?

**Comment by maxkatz6** [comment]
DragDrop.DoDragDrop for that. But you need to do it from pointer event handler, like pointer pressed.

**Reply by followLemmi** [reply]
```
private async void FileList_OnCellPointerPressed(object? sender, PointerPressedEventArgs e)
{
        var fileModel = (e.Source as Control)!.DataContext as FileModel;
        var dataObject = new DataObject();
        var filePaths = new[] { fileModel.FullPath };
        Console.WriteLine($"Files to drop {fileModel.FullPath}");
        dataObject.Set(DataFormats.Files, filePaths);
        
        
        var dragDrop = await DragDrop.DoDragDrop(e, dataObject, DragDropEffects.Copy);
}
```
If I drop file to Windows desktop from Avalonia app like this, Windows freeze about few seconds and after reload self UI. File not dropped to  desktop. I pretty shure that i doing something wrong, but i cant find any solution

**Reply by maxkatz6** [reply]
DataFormats.Files doesn't expect file paths, but `IEnumerable<IStorageItem>`.
If you want to drop file paths, use DataFormats.FileNames.

**Reply by followLemmi** [reply]
Which type of IEnumerable<?> DataFormats.Files expected for?

**Reply by maxkatz6** [reply, marked as answer]
Sorry, github ate my generics... `IEnumerable<IStorageItem>` it is.

**Reply by maxkatz6** [reply]
Which can be obtained using StorageProvider API. It is necessary in sandboxed environments, where application don't have plain access to the file system. Like, browser, mobile or app-store macOS apps.

**Reply by followLemmi** [reply]
Many thanks to you=)

**Reply by followLemmi** [reply]
@maxkatz6 does this solution work for MacOS? When I launched the same working solution on MacOS, the DnD event is initiated, but when I try to transfer it, for example, to Finder, nothing happens

**Reply by laonzh** [reply]
same too.

**Reply by jamsoft** [reply]
I'm seeing the same issue on macOS. My app has hardened runtime enabled and is signed and notarized. I also have the com.apple.security.automation.apple-events entitlement but Dragging from my avalonia app to finder still does nothing.

This is all working fine on Windows (as I understand it the Linux X11 protocol still isn't implemented in 11.x.x)

**Reply by maxkatz6** [reply]
@jamsoft yes, on macOS hardened runtime you would have to use IStorageFile APIs.
But at current master it won't work well, as it requires https://github.com/AvaloniaUI/Avalonia/pull/16090 to be merged. + dragndrop specific code needs to be adjusted on macOS too, which wasn't yet done.

**Reply by maxkatz6** [reply]
Though, https://github.com/AvaloniaUI/Avalonia/pull/16090 alone might be enough to get it work well on hardened runtime.

**Reply by jamsoft** [reply]
Thanks for the reply! I thought I was going mad.

I have my app all signed and notarized with hardened runtime and using IStorageFile APIs already. Works great on Windows and nowhere else.

Thanks for confirming.

**Reply by maxkatz6** [reply]
If you can test that PR builds, see https://github.com/AvaloniaUI/Avalonia/pull/16090#issuecomment-2184311766, it might help greatly.
Is it possible to test hardened runtime and a sandboxing without uploading app to the store?

**Reply by workgroupengineering** [reply]
https://stackoverflow.com/questions/24947661/os-x-app-testing-for-sandbox-violations

**Reply by jamsoft** [reply]
> If you can test that PR builds, see [#16090 (comment)](https://github.com/AvaloniaUI/Avalonia/pull/16090#issuecomment-2184311766), it might help greatly. Is it possible to test hardened runtime and a sandboxing without uploading app to the store?

I distribute my macOS application on my own web site, not via the App Store. I'll have a look at testing that PR build but it would take a while to drop this into my application.

**Comment by Bodekaer** [comment]
> DragDrop.DoDragDrop for that. But you need to do it from pointer event handler, like pointer pressed. 

I've been trying this, however the MacOS accepts only Text and not Files data format. Basically no matter which file of IStorage array/list I provide, it's like the MacOS just ignores/doesn't accept it.

Can anyone provide a quick code example or documentation link? @maxkatz6 perhaps?

**Comment by pawn86** [comment]
Hello, I'm a beginner with Avalonia. I've been trying to implement file drag-and-drop to Windows 10 File Explorer, but it's not working—no files appear at the drop target.Can you help me. Can’t wait to hear back from you!
<img width="1253" height="729" alt="image" src="https://github.com/user-attachments/assets/d3377902-4bce-4206-aacf-04f558e4197d" />

**Reply by timunie** [reply]
Have you set AllowDrop in the XAML and have you made sure that your Main void in Program.cs is [STAThread] annotated?

**Reply by pawn86** [reply]
> Have you set AllowDrop in the XAML and have you made sure that your Main void in Program.cs is [STAThread] annotated?
Hi, Thank you reply. Allow Drop is not recognized in DataGrid.
    [STAThread]
    public static void Main(string[] args) => BuildAvaloniaApp()
        .StartWithClassicDesktopLifetime(args);

<img width="1045" height="597" alt="image" src="https://github.com/user-attachments/assets/4388954d-3017-4603-b744-a64995103d0d" />

**Reply by timunie** [reply]
It's probably an attached property. Try DragDrop.AllowDrop

**Reply by pawn86** [reply]
> It's probably an attached property. Try DragDrop.AllowDrop

    <DataGrid
        AreRowDetailsFrozen="True"
        Background="Gray"
        ColumnHeaderHeight="35"
        DragDrop.AllowDrop="True"
        GridLinesVisibility="All"
        HorizontalGridLinesBrush="DodgerBlue"
        ItemsSource="{Binding NotifyView}"
        PointerMoved="InputElement_OnPointerMoved"
        VerticalGridLinesBrush="DodgerBlue">
        <DataGrid.Columns>
            <DataGridTextColumn
                Width="100"
                Binding="{Binding Id}"
                Header="编号" />
            <DataGridTextColumn
                Width="*"
                Binding="{Binding Path}"
                Header="路径" />
        </DataGrid.Columns>
    </DataGrid>  It‘s not working. I don't know why. Could you help me?
